src/NServiceBus.AmazonSQS/MessagePump.cs

```diff
@@ -59,220 +58,257 @@ public MessagePump(ConnectionConfiguration configuration, IAmazonS3 s3Client, IA
+            maxConcurrency = limitations.MaxConcurrency;
 
-            for (var i = 0; i < concurrencyLevel; i++)
+            int numberOfPumps;
+            if (maxConcurrency <= 10)
             {
-                consumerTasks.Add(ConsumeMessages());
+                numberOfPumps = 1;
+                numberOfMessagesToFetch = maxConcurrency;
+            }
+            else
+            {
+                numberOfMessagesToFetch = 10;
+                numberOfPumps = Convert.ToInt32(Math.Ceiling(Convert.ToDouble(maxConcurrency) / numberOfMessagesToFetch));
+            }
+
+            receiveMessagesRequest = new ReceiveMessageRequest
+            {
+                MaxNumberOfMessages = numberOfMessagesToFetch,
+                QueueUrl = queueUrl,
+                WaitTimeSeconds = 20,
+                AttributeNames = new List<string>
+                {
+                    "SentTimestamp"
+                }
+            };
+
+            maxConcurrencySempahore = new SemaphoreSlim(maxConcurrency);
+            pumpTasks = new List<Task>(numberOfPumps);
+
+            for (var i = 0; i < numberOfPumps; i++)
+            {
+                pumpTasks.Add(Task.Run(() => ConsumeMessages(cancellationTokenSource.Token), CancellationToken.None));
             }
         }
 
         public async Task Stop()
         {
             cancellationTokenSource?.Cancel();
 
+            await Task.WhenAll(pumpTasks).ConfigureAwait(false);
+
+            pumpTasks?.Clear();
+            cancellationTokenSource?.Dispose();
+            maxConcurrencySempahore?.Dispose();
+        }
+
+        async Task ConsumeMessages(CancellationToken token)
+        {
+            // cached and reused per receive loop
+            var concurrentReceiveOperations = new List<Task>(numberOfMessagesToFetch);
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    var receivedMessages = await sqsClient.ReceiveMessageAsync(receiveMessagesRequest, token).ConfigureAwait(false);
+
+                    ProcessMessages(receivedMessages.Messages, concurrentReceiveOperations, token);
+
+                    await Task.WhenAll(concurrentReceiveOperations).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // ignore for graceful shutdown
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error("Exception thrown when consuming messages", ex);
+                }
+                finally
+                {
+                    concurrentReceiveOperations.Clear();
+                }
+            } // while
+        }
+
+        // ReSharper disable once ParameterTypeCanBeEnumerable.Local
+        // ReSharper disable once SuggestBaseTypeForParameter
+        void ProcessMessages(List<Message> receivedMessages, List<Task> concurrentReceiveOperations, CancellationToken token)
```

Comment:
Can we change `List<Message> receivedMessages` to `IEnumerable<Message> receivedMessages`?

Reply:
Nope this would mean allocs